core/vm/evm.go

```diff
@@ -139,21 +142,31 @@ type EVM struct {
 	// applied in opCall*.
 	callGasTemp uint64
 
-	ETXCache     []*types.Transaction
-	ETXCacheLock sync.RWMutex
+	ETXCache              []*types.Transaction
+	CoinbaseDeletedHashes []*common.Hash
+	CoinbasesDeleted      map[[47]byte][]byte
+	ETXCacheLock          sync.RWMutex
+	Batch                 ethdb.Batch
 }
 
 // NewEVM returns a new EVM. The returned EVM is not thread safe and should
 // only ever be used *once*.
-func NewEVM(blockCtx BlockContext, txCtx TxContext, statedb StateDB, chainConfig *params.ChainConfig, config Config) *EVM {
+func NewEVM(blockCtx BlockContext, txCtx TxContext, statedb StateDB, chainConfig *params.ChainConfig, config Config, batch ethdb.Batch) *EVM {
 	evm := &EVM{
-		Context:     blockCtx,
-		TxContext:   txCtx,
-		StateDB:     statedb,
-		Config:      config,
-		chainConfig: chainConfig,
-		chainRules:  chainConfig.Rules(blockCtx.BlockNumber),
-		ETXCache:    make([]*types.Transaction, 0),
+		Context:               blockCtx,
+		TxContext:             txCtx,
+		StateDB:               statedb,
+		Config:                config,
+		chainConfig:           chainConfig,
+		chainRules:            chainConfig.Rules(blockCtx.BlockNumber),
+		ETXCache:              make([]*types.Transaction, 0),
+		CoinbaseDeletedHashes: make([]*common.Hash, 0),
+		CoinbasesDeleted:      make(map[[47]byte][]byte),
+	}
+	if batch != nil {
+		evm.Batch = batch
+	} else {
+		evm.Batch = memorydb.New(log.Global).NewBatch() // Just used as a cache for simulating calls
```

Comment:
memory db batch does not implement GetPending functionality, so we should either force the level db batch or if it can be nil, implement the GetPending for the memorydb

Reply:
Why does it need GetPending for simulating the EVM?